### PR TITLE
PWX-24976 : Add schema definition for resource transformation CR

### PR DESCRIPTION
**Pull request by ram-infrac**
**What type of PR is this?**
> feature

**What this PR does / why we need it**:
This PR adds -
-  schema definition for new `ResourceTransformation` CR
-  Fixes issue where FA/FB volume picked up for migration

**Does this PR change a user-facing CRD or CLI?**:
yes, new ResourceTransformation CR is introduced.

**Is a release note needed?**:
yes

**Does this change need to be cherry-picked to a release branch?**:
2.12

**Note**: Will add CR change, release note section in final PR for this feature

### CODE REVIEW

**Review by lpitstickpstg — APPROVED** [review]

File: drivers/volume/portworx/portworx.go [hunk 1]
```diff
@@ -2421,7 +2421,7 @@ func (p *portworx) StartMigration(migration *storkapi.Migration) ([]*storkapi.Mi
 			return nil, fmt.Errorf("error getting list of volumes to migrate: %v", err)
 		}
 		for _, pvc := range pvcList.Items {
-			if !p.OwnsPVC(core.Instance(), &pvc) {
+			if !p.IsSupportedPVC(core.Instance(), &pvc, true) {
```
Comment (adityadani) [inline drivers/volume/portworx/portworx.go:2424]: nit: is this part of a different bug / PR ? How is this related to resource transformation?

File: pkg/apis/stork/v1alpha1/resourcetransformation.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+type ResourceTransformationOperationType string
+
+const (
+	// AddPathValue is used to add path+value in specified resource spec
+	// if path+value already exist this operation will override value
+	// at given path
+	AddResourcePath ResourceTransformationOperationType = "add"
+	// ModifyPathValue is used to merge value at speficied resource path
+	// in case slice, entries will be appended.
+	// in case keypair, entries will be merged
```
Comment (adityadani) [inline pkg/apis/stork/v1alpha1/resourcetransformation.go:25]:
```suggestion
	// in case of a  slice, entries will be appended.
	// in case of a key pair, entries will be merged
```

File: pkg/apis/stork/v1alpha1/resourcetransformation.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	ModifyPathValue ResourceTransformationOperationType = "modify"
+	// DeletePath from resource specification
+	DeleteResourcePath ResourceTransformationOperationType = "delete"
+	// JsonResourcePatch will path json in given resource spec
```
Comment (adityadani) [inline pkg/apis/stork/v1alpha1/resourcetransformation.go:29]:
```suggestion
	// JsonResourcePatch will patch json in given resource spec
```

File: pkg/apis/stork/v1alpha1/resourcetransformation.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	KeyPairResourceType ResourceTransformationValueType = "keypair"
+)
+
+// ResourceTransformationStatsusType is status of resource transformation CR
+type ResourceTransformationStatusType string
+
+const (
+	// ResourceTransformationStatusInitial represents initial state of resource
+	// transformation CR
+	ResourceTransformationStatusInitial ResourceTransformationStatusType = "int"
```
Comment (adityadani) [inline pkg/apis/stork/v1alpha1/resourcetransformation.go:56]: int -> initial
int - in-progress

File: pkg/apis/stork/v1alpha1/resourcetransformation.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+type ResourceTransformationSpec struct {
+	Objects []TransformSpecs `json:"transformSpecs"`
+}
+
+// TransformSpecs specifies the patch to update selected resource
+// before migration/restore
+type TransformSpecs struct {
+	// Selectors label selector to filter out resource for
+	// patching
+	Selectors map[string]string `json:"selectors"`
```
Comment (adityadani) [inline pkg/apis/stork/v1alpha1/resourcetransformation.go:90]: Should we take an object type here as well (GVK)? A complete application stack involving a PVC, Service, Deployment, ConfigMap can potentially have the same selectors, but we might want a transformation only on a Service object. And the type can take a wildcard `*` which would allow us to apply to all the resources if required.

**Review by adityadani — CHANGES_REQUESTED** [review]

File: drivers/volume/portworx/portworx.go (hunk 1 above)
Comment (ram-infrac) [reply drivers/volume/portworx/portworx.go:2424]: yes included this included as separate commit  https://github.com/libopenstorage/stork/pull/1130/commits/75b4a674f7de72e498c4ea2a0d6b671939046fb4
